lib/ruby_lsp/queue.rb

```diff
@@ -0,0 +1,189 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "benchmark"
+
+module RubyLsp
+  class Queue
+    class Cancelled < Interrupt; end
+
+    extend T::Sig
+
+    class Result < T::Struct
+      const :response, T.untyped # rubocop:disable Sorbet/ForbidUntypedStructProps
+      const :error, T.nilable(StandardError)
+      const :request_time, T.nilable(Float)
+    end
+
+    class Job < T::Struct
+      extend T::Sig
+
+      const :request, T::Hash[Symbol, T.untyped]
+      prop :cancelled, T::Boolean
+
+      sig { void }
+      def cancel
+        self.cancelled = true
+      end
+    end
+
+    sig do
+      params(
+        writer: LanguageServer::Protocol::Transport::Stdio::Writer,
+        handlers: T::Hash[String, Handler::RequestHandler]
+      ).void
+    end
+    def initialize(writer, handlers)
+      @writer = writer
+      @handlers = handlers
+      # The job queue is the actual list of requests we have to process
+      @job_queue = T.let(Thread::Queue.new, Thread::Queue)
+      # The jobs hash is just a way of keeping a handle to jobs based on the request ID, so we can cancel them
+      @jobs = T.let({}, T::Hash[T.any(String, Integer), Job])
```

Comment:
Can we enforce job ids to have the same type? So we don't need to handle 2 types of them.

Reply:
I think the LSP interface allows them to be either?

Reply:
The [specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#cancelRequest) mentions ID can be a string or integer, so I figured it would be better to match it.

Reply:
I see. Yeah although it's unlikely but there could be both `1` and `"1"` being sent.